src/sources/redis/mod.rs

```diff
@@ -167,34 +172,56 @@ impl SourceConfig for RedisSourceConfig {
         match self.data_type {
             DataTypeConfig::List => {
                 let list = self.list.unwrap_or_default();
-                list::watch(
+                list::watch(list::WatchInputs {
                     client,
-                    bytes_received.clone(),
-                    self.key.clone(),
-                    self.redis_key.clone(),
-                    list.method,
+                    bytes_received: bytes_received.clone(),
+                    key: self.key.clone(),
+                    redis_key: self.redis_key.clone(),
+                    method: list.method,
                     decoder,
                     cx,
-                )
+                    log_namespace,
+                })
                 .await
             }
             DataTypeConfig::Channel => {
-                channel::subscribe(
+                channel::subscribe(channel::SubscribeInputs {
                     client,
                     connection_info,
-                    bytes_received.clone(),
-                    self.key.clone(),
-                    self.redis_key.clone(),
+                    bytes_received: bytes_received.clone(),
+                    key: self.key.clone(),
+                    redis_key: self.redis_key.clone(),
                     decoder,
                     cx,
-                )
+                    log_namespace,
+                })
                 .await
             }
         }
     }
 
-    fn outputs(&self, _global_log_namespace: LogNamespace) -> Vec<Output> {
-        vec![Output::default(self.decoding.output_type())]
+    fn outputs(&self, global_log_namespace: LogNamespace) -> Vec<Output> {
+        let log_namespace = global_log_namespace.merge(self.log_namespace);
+
+        let schema_definition = self.decoding.schema_definition(log_namespace);
+
+        let redis_key_path = self
```

Comment:
if you move `redis_key_path` a few lines up, all of the `schema_definitions` can be collapsed into the "builder pattern" so you only need to assign it once.